### PR TITLE
Speed up MAST queries

### LINKED ISSUE #941: Slow/incomplete results for NIRCam from MAST API search
Recent timing tests on the test server showed that when using the mast_query_filenames_by_instrument() function, it took upwards of 3 minutes to get a response for NIRCam, while it took under 1 second for the other instruments. I also saw that the NIRCam query was returning a list of exactly 50,000 filenames, which seemed suspicious.

I submitted a ticket to the Archive help desk about this. The here is the response:

```
"The behavior you're seeing is expected. There is a default "pagesize" set in "Mast.service_request_async"
that limits the returned results to 50,000 files. Because there are more than 50,000 NIRCAM files in the JWST
Instrument Keyword database (51,796 files at the time I'm writing this message), your query is being slowed
down significantly and you're only getting part of the expected result.

Your best option will likely be to break your single query up into multiple queries using the "filters" parameter
to search by e.g., program ID or date."
```

So, it seems like we will have to rethink how to query the archive, at least within mast_query_filenames_by_instrument(). Do we really need all of the filenames? Perhaps we can filter by suffix, and just get the e.g. uncal files? Or maybe we can query to get a list of program IDs, and then query for each?

A good place to start may be to check where we call this function and see what we really end up using from the results.

### PR DESCRIPTION
**Pull request by bhilbert4**
Resolves #941 

This PR updates JWQL's MAST queries to avoid issues caused by query results that are too large.  According to the Archive Branch, query results are limited to 50,000 rows. Any query result that is larger is cropped down to 50,000 rows and results in a much slower response. When querying for all files for a particular instrument, as we are set up to do in multiple functions in JWQL, we are already beyond this 50,000 row limit for NIRCam. So a NIRCam search was taking upwards of 3 minutes, while a search with one of the other instruments was taking under 1 second.

The easiest way around this I think is to search per proposal. So this PR re-works our MAST queries to run one proposal at a time.

NOTE: MAST queries can be filtered using essentially any of the JWST keywords. Using proposal ID seemed the easiest solution to me, but there may be others. Here is the list of possible filters: https://mast.stsci.edu/api/v0/_jwst_inst_keywd.html

### CODE REVIEW

File: jwql/website/apps/jwql/data_containers.py [hunk 1]
```diff
@@ -208,13 +208,10 @@ def get_all_proposals():
...
         filesystem
     """
-
-    proposals = glob.glob(os.path.join(FILESYSTEM_DIR, 'public', '*'))
-    proposals.extend(glob.glob(os.path.join(FILESYSTEM_DIR, 'proprietary', '*')))
-    proposals = sorted(list(set(proposals)))
-    proposals = [proposal.split('jw')[-1] for proposal in proposals]
-    proposals = [proposal for proposal in proposals if len(proposal) == 5]
-
+    proprietary_proposals = os.listdir(os.path.join(FILESYSTEM_DIR, 'proprietary'))
```
Comment (bhilbert4) [inline jwql/website/apps/jwql/data_containers.py:212]: In my limited speed tests, os.listdir was faster than glob

**Comment by bhilbert4** [comment]
- [x] get_expstart is quite expensive. ~0.1 second for each call, and it's called on all files in a proposal. Better to include expstart column in the mast query for the filenames.

**Comment by bhilbert4** [comment]
data_containers.get_thumbnails_all_instruments() needs to be revised in a similar way. That is a more complex query, so might take some more thought. Maybe looping over proposal numbers would work. I've split that off into a separate issue, #943

**Comment by bhilbert4** [comment]
At this point, it looks like the web app is faster in some cases. When loading a page for a particular proposal, it's quite fast for proposals that have under roughly 150 files. But for proposals with more, the loading time is still very long, if it loads at all.

I've run thumbnails_ajax() manually on the server, and even for proposals with 1800 files, the run time is under 1 second. So the slowdown must be coming in somewhere else. The javascript functions? The slowdown is strange. It's going from a reasonably quick load to never loading the page as the number of files increases. There doesn't seem to be a middle ground where it takes e.g. 60 seconds. @mfixstsci any thoughts? 

I also manually ran views.archived_proposals_ajax() on the server, and that seemed reasonably quick as well. Loading time for each instrument's Archive page now takes ~20-25 seconds. The bottleneck here is querying MAST once for each proposal. I'm not sure how else to break up the MAST queries in order to assure we don't hit the 50,000 image limit and still cover all the data. Perhaps query over some time period, but then we'd have to organize the results by proposal.

**Comment by bhilbert4** [comment]
While looking at lists of thumbnail files, I was able to find and resolve #801. For a given instrument, the entries in the archive page now report the same number of files as on the thumbnail pages (from my spot checks). This means that #821 can be closed without merging.

**Comment by bhilbert4** [comment]
The only remaining issue at the moment is that the thumbnail pages for proposals containing more than ~300 files do not load due to what looks to be a server timeout error. (net::ERR_CONTENT_LENGTH_MISMATCH 200). From my reading about it, it sounds like this may be because a json file is too large, or there is some delay during the reading in of that file. See [here](https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_server_keepalivetimeout) and [here](https://www.geeksforgeeks.org/node-js-http-server-keepalivetimeout-property/). @mfixstsci do you know anything about this? 

It sounds like a possible fix would be an adjustment of the keepAliveTimeout value on the server. This may also be an argument in favor of tweaking the way we display proposals that have lots of files. In some (rare) cases for NIRCam there are proposals with thousands of files. It might be easier for users to navigate in cases like these if we organize the thumbnails and display only those for one Observation at a time.

**Comment by bhilbert4** [comment]
The previous commit addresses #802. Program IDs are now in descending order when the Archive page is loaded for a particular instrument. The thumbnails for each program's page are sorted into reverse chronological order. Would be nice to sort the exposures into reverse chronological order, and within each exposure, sort by detector, if possible. That might take some work in data_containers.thumbnails_ajax()

File: environment_python_3_7_with_bokeh2.1.1.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+channels:
```
Comment (mfixstsci) [inline environment_python_3_7_with_bokeh2.1.1.yml]: @bhilbert4 did you mean to add this file?

File: jwql/jwql_monitors/monitor_mast.py [hunk 3]
```diff
@@ -143,8 +143,13 @@ def instrument_keywords(instrument, caom=False):
         A DataFrame of the keywords
     """
     # Retrieve one dataset to get header keywords
+    if not caom:
```
Comment (mfixstsci) [inline jwql/jwql_monitors/monitor_mast.py:146]: What does this logic do? I am just curious, what is special about this program?

File: jwql/website/apps/jwql/static/js/jwql.js [hunk 4]
```diff
@@ -520,10 +520,19 @@ function update_thumbnails_page(inst, proposal, base_url) {
         url: base_url + '/ajax/' + inst + '/archive/' + proposal + '/',
         success: function(data){
             // Perform various updates to divs
+            let t1 = performance.now();
```
Comment (bhilbert4) [inline jwql/website/apps/jwql/static/js/jwql.js]: Oops. These timing statements can be deleted. I'll do that.

Comment (mfixstsci) [inline jwql/website/apps/jwql/static/js/jwql.js]: Are we logging this information now or was this for testing the loading speeds?

File: jwql/website/apps/jwql/views.py [hunk 5]
```diff
@@ -271,36 +271,61 @@ def archived_proposals_ajax(request, inst):
...
-        try:
-            relative_filepath = filesystem_path(filename, check_existence=False)
-            full_filepath = os.path.join(filesystem, 'proprietary', relative_filepath)
-            filenames.append(full_filepath)
-        except ValueError:
-            print('Unable to determine filepath for {}'.format(filename))
-
-    # Gather information about the proposals for the given instrument
-    proposal_info = get_proposal_info(filenames)
+    for proposal in all_proposals:
```
Comment (mfixstsci) [inline jwql/website/apps/jwql/views.py:282]: This is just where we are breaking the views into proposals based on instrument and propid? This looks more efficient 👍

**Review by mfixstsci — APPROVED** [review]
Looks good, thanks for submitting @bhilbert4 !

File: jwql/jwql_monitors/monitor_mast.py (hunk 3 above)
Comment (bhilbert4) [reply jwql/jwql_monitors/monitor_mast.py:146]: This block queries two different MAST services. Either Caom.Filtered or Jwst.Instrument.Filtered.<instrument>. The two services seem to use two different keywords for the program ID, so this just makes sure the correct keyword is used.

File: jwql/website/apps/jwql/views.py (hunk 5 above)
Comment (bhilbert4) [reply jwql/website/apps/jwql/views.py:282]: Yes, I like this much better. This pushes more of the filtering work onto MAST, which is faster than when we do it ourselves.

File: environment_python_3_7_with_bokeh2.1.1.yml (hunk 2 above)
Comment (bhilbert4) [reply environment_python_3_7_with_bokeh2.1.1.yml]: Removed. Not sure how it got in there.

File: jwql/website/apps/jwql/static/js/jwql.js (hunk 4 above)
Comment (bhilbert4) [reply jwql/website/apps/jwql/static/js/jwql.js]: Leftover from testing. Removed. I am excited that I now know how to print information to the console from javascript functions though.

**Pull request merged**